### DISCUSSION TITLE
Hide parameters with CallerXxxAttribute from interfaces

### DISCUSSION THREAD

**Discussion by ENikS** [dotnet/csharplang#4950] [Language Ideas]
I am writing a SourceGenerator for [dependency injection](https://github.com/unitycontainer) container. 
In the generator I am visiting the code tree and recording occurrences of Register and Resolve method calls. Based on analysis of the calls I am generating implementation for these methods. 

To match these implementations to actual method calls at runtime I need some form of ID. If I add something like:
```c#
public object Resolve(some parameter,
        [CallerMemberName] string memberName = "",
        [CallerFilePath] string sourceFilePath = "",
        [CallerLineNumber] int sourceLineNumber = 0)
```  
it will provide all that I need, but I would have to add it to the public API, defined by interface:

```c#
public interface SomeInterface
{
    public object Resolve(some parameter)
}
```  
I wonder if these compiler generated parameters could be added to implementation without adding them to the public API? From interface point of view these parameters are irrelevant and only meant for the compiler.

Or, perhaps there is a way to hide these from a consumer?

**Comment by CyrusNajmabadi** [comment]
> Or, perhaps there is a way to hide these from consumer?

No, there is no way to do this.  The consumer is the one that passes these values in.  They only know to do this by being able to see and tell that those parameters are there so that it can fill them in at the callsite with teh appropriate values.

**Reply by ENikS** [reply]
When I say consumer, I mean people who use the API

**Reply by 333fred** [reply]
So does Cyrus. These need to be visible so the call-site (ie, the consumer you are both talking about) can pass the values.

**Reply by ENikS** [reply]
Perhaps `CallerXxxAttribute` could use a property:

```c#
public interface SomeInterface
{
    public object Resolve(some parameter,
        [CallerMemberName] string memberName = "",
        [CallerFilePath(CompilerOnly = true)] string sourceFilePath = "",
        [CallerLineNumber(CompilerOnly = true)] int sourceLineNumber = 0)
}
```

**Reply by CyrusNajmabadi** [reply]
All CallerXXXAttributes are effectively 'compiler only'.  They exist to let the compiler know that these values need to be passed in at the callsite.  The callsite must know these attributes are there so that it can actually substitute tehse values in at compile time.

**Comment by svick** [comment]
The compiler needs those parameters, so they have to be in the public API. But since the programmer who's calling the method generally doesn't need to know about them, maybe the IDE could do something to hide those parameters in IntelliSense, or at least make them less prominent?

**Reply by CyrusNajmabadi** [reply]
Sure.  That's a request though for dotnet/roslyn and how the IDE 'signature help/quickinfo/etc.' features work :)

**Comment by ENikS** [comment]
I will ask this question in Roslyn repo. Thank you for your time.
